ws/src/main/java/org/kaazing/netx/ws/internal/io/WsMessageWriter.java

```diff
@@ -16,43 +16,252 @@
+                    connection.getOutputStream().writeBinary(CONTINUATION, binaryBuffer, 0, binaryBufferOffset, true);
+                }
+                initialFrame = true;
+                binaryBufferOffset = 0;
+            }
+            finally {
+                lock.unlock();
+            }
+        }
+
+        @Override
+        public void close() throws IOException {
+            flush();
+        }
+    }
+
+    private static class WsTextWriter extends Writer {
+        private static final String MSG_INDEX_OUT_OF_BOUNDS = "offset = %d; (offset + length) = %d; buffer length = %d";
+
+        private final WsURLConnectionImpl connection;
+        private final char[] textBuffer;
+        private final Lock lock;
+
+        private int textBufferOffset;
+        private boolean initialFrame;
+
+        public WsTextWriter(WsURLConnectionImpl connection) {
+            this.connection = connection;
+            this.textBuffer = new char[connection.getMaxFramePayloadLength()];
+            this.lock = new OptimisticReentrantLock();
+            this.initialFrame = true;
+            this.textBufferOffset = 0;
+        }
+
+        @Override
+        public void write(char[] cbuf, int offset, int length) throws IOException {
+            if (cbuf == null) {
+                throw new NullPointerException("Null buffer passed in");
+            }
+            else if ((offset < 0) || (length < 0) || (offset + length > cbuf.length)) {
+                throw new IndexOutOfBoundsException(format(MSG_INDEX_OUT_OF_BOUNDS, offset, offset + length, cbuf.length));
+            }
+
+            try {
+                lock.lock();
+
+                while (length > 0) {
+                    int len = Math.min(length, textBuffer.length - textBufferOffset);
+
+                    System.arraycopy(cbuf, offset, textBuffer, textBufferOffset, len);
+                    textBufferOffset += len;
+
+                    if (textBufferOffset == textBuffer.length) {
+                        Opcode opcode = initialFrame ? TEXT : CONTINUATION;
+                        connection.getWriter().writeText(opcode, textBuffer, 0, textBuffer.length, false);
+                        initialFrame = false;
+                        textBufferOffset = 0;
+                    }
+
+                    offset += len;
+                    length -= len;
+                }
+            }
+            finally {
+                lock.unlock();
+            }
+        }
+
+        @Override
+        public void flush() throws IOException {
+            try {
+                lock.lock();
+
+                if (initialFrame) {
+                    // Only one frame in the message.
+                    connection.getWriter().write(textBuffer, 0, textBufferOffset);
+                }
+                else {
+                    // Send the final frame.
+                    connection.getWriter().writeText(CONTINUATION, textBuffer, 0, textBufferOffset, true);
```

Comment:
`FIN` bit should be set on `close()`, not `flush()`, right?
Of course, `close()` may trigger a partial final frame.  `flush()` may also trigger a partial non-final frame.